[FEAT] Metronome 컴포넌트 추가

nakisara01 opened the pull request:
### 1. 🚀 어떤 종류의 PR인가요? (What type of PR is this?)
- [ ] **Bug**: 예상과 다르게 동작하는 문제 또는 버그를 수정합니다.
- [ ] **Chore**: 빌드, 설정, 라이브러리 업데이트 등 사용자에게 직접적인 영향이 없는 작업을 합니다.
- [ ] **Documentation**: 문서(README, Wiki, API 문서 등)를 추가하거나 수정합니다.
- [x] **Feature**: 새로운 기능을 추가합니다.
- [ ] **Refactor**: 버그 수정이나 기능 추가 없이 기존 코드를 개선합니다.
- [ ] **Style**: 코드 포맷팅, 세미콜론 누락 등 코드 스타일에 관련된 것을 수정합니다.
- [ ] **Test**: 테스트 코드를 추가하거나 수정합니다.


### 2. 📝 개요 (Description)
➡️ *가이드: PR의 목적을 한두 문장으로 요약합니다. '무엇을' & '왜' 수정했는지 리뷰어가 쉽게 이해할 수 있도록 작성하는 것이 핵심입니다.*
Class Diagram 양식에 맞춘 Metronome Component 개발

### 3. 🔗 관련 이슈 (Related Issue)
➡️ *가이드: 관련된 이슈 번호를 `Closes #이슈번호` 형식으로 적어주세요. 이렇게 하면 PR이 머지될 때 해당 이슈가 자동으로 종료되어 편리합니다.*
Closes #5 


### 4. ✅ 작업 목록 (Checklist
➡️ *가이드: 리뷰어가 코드 변경 사항을 쉽게 따라갈 수 있도록, 실제 작업 내역을 구체적인 단위로 나누어 나열합니다.*
- Metronome(bpm: , isPlaying: ,selectedTimeSig: ) 형태로 Metronome 컴포넌트를 사용할 수 있습니다.
- 메트로놈에서 강박, 약박이 나뉘어 소리가 플레이됩니다.
- 메트로놈 3/4 박자, 4/4 박자, 6/8 박자에 따른 컴포넌트 UI 갯수가 분리됩니다.

### 5. 📸 스크린샷 (Screenshots)
➡️ *가이드: UI 변경이 있다면, 수정 전(Before)과 후(After)의 모습을 스크린샷으로 첨부하여 리뷰어가 직관적으로 변화를 이해하도록 돕습니다.*
|<img width="200" height="600" alt="스크린샷 2025-10-22 오후 5 18 37" src="https://github.com/user-attachments/assets/97fc3277-218a-4b36-b46b-c437fdf6bfe1" />|<img width="200" height="600" alt="스크린샷 2025-10-22 오후 5 19 12" src="https://github.com/user-attachments/assets/91a71440-ac97-4658-86d8-8aec43c62f99" />|<img width="200" height="600" alt="스크린샷 2025-10-22 오후 5 19 32" src="https://github.com/user-attachments/assets/ffb0e329-0e70-4bd5-909b-8a28071940fe" />|



### 6. ❗ 리뷰어에게 전달할 기타 사항 (Other information for reviewers)
➡️ *가이드: 리뷰어가 놓치지 않고 꼭 확인해야 할 핵심 코드 위치나, 특정 방식으로 테스트가 필요한 경우 친절하게 안내합니다. 리뷰 효율을 크게 높일 수 있습니다.*
- Metronome body 내에서 .animation, .onChange 등의 로직이 많아 뷰를 보기 어렵다고 느껴집니다. 혹시 이 부분을 따로 함수로 빼서 넣을 수 있는 방안이 있다면 제안 부탁드립니다.
- MetronomeControlView는 간단하게 설정을 변경하며 Metronome 컴포넌트를 테스트할 수 있는 뷰입니다. 추후 삭제할 예정입니다.
- Metronome 내에서 BPM, TimeSignature 을 선언한 모델을 반영했습니다. **정신이 온전치 않은 상태**에서 수정한 사항이어서 제대로 사용했는지 확인 한 번 부탁드립니다.

nakisara01 committed:
[feat] Add metronome component

nakisara01 committed:
Merge remote-tracking branch 'origin/dev' into feat/#5

nakisara01 committed:
[feat] Apply Class diagram structure in Metronome

- 선언된 TimeSignature 이용하도록 로직 조정
- cd_empty/fill image 추가
- Metronome Sound 추가
- 원래 존재하던 TimeSignature 삭제

nakisara01 committed:
[feat] Add Metronome Use Example & Annotation

Segyun commented on GMG/GMG/Core/Models/MusicNotation/TimeSignature.swift:
```diff
@@ -1,6 +1,6 @@
 //  Copyright © 2025 ADA 4th GMG. All rights reserved.
 
-struct TimeSignature {
+struct TimeSignature: Equatable {
```
extension으로 빼는 것 어떨까요?

```suggestion
extension TimeSignature: Equatable {}
```

nakisara01 replied:
수정했습니다!

Segyun commented on GMG/GMG/Resources/Assets.xcassets/MetronomeCD/metronome_cd_empty.imageset/Contents.json:
`MetronomeCDEmpty`로 CamelCase로 통일하는 것이 어떨까요?

Audios의 메트로놈 소리 파일도 수정하면 좋을 것 같습니다!

nakisara01 replied:
디자인 변경되면서 Empty, Fill의 구분이 사라져서 MetronomeCD로 변경하여 삽입했습니다, 음원 파일도 MetronomeHigh, MetronomeLow로 변경했습니다.

ehdrbdndns commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,149 @@
+        }
+    }
+
+    // 위에서 구분한 Index에 따라 박자별 High or Low 음성 재생
```
```suggestion
    // 박자별 High or Low 음성 재생
```
위 주석이 더 적합할 것 같스비당

nakisara01 replied:
적용했습니다

ehdrbdndns commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,149 @@
+                    lastStepPlayed = step
+                    playClick(for: currentBeatToShow)
+                }
+            }
```
modifier 들여쓰기 넣어주면 더 읽기 편할 것 같아요!
```suggestion
            BeatBar(beatsPerBar: beatsPerBar, currentBeat: currentBeatToShow)
                .animation(.spring(response: 0.28, dampingFraction: 0.85, blendDuration: 0.2), value: beatsPerBar)
                .padding()
                .onChange(of: context.date) { _, _ in
                    guard isPlaying, let start = startDate else { return }
                    let step = Int(floor(Date().timeIntervalSince(start) / beatInterval))
                    if step != lastStepPlayed {
                        lastStepPlayed = step
                        playClick(for: currentBeatToShow)
                    }
                }
```

ehdrbdndns commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,149 @@
+    }
+
+    // 위에서 구분한 Index에 따라 박자별 High or Low 음성 재생
+    private func playClick(for beatIndex: Int) {
```
playBeatSound()가 더 직관적인듯??

ehdrbdndns commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,149 @@
+        }
+        .onChange(of: bpm.value) { _, _ in
+            if isPlaying { startDate = Date(); lastStepPlayed = -1 }
+        }
```
오디오 관련 메모리 해제 로직이 필요해보입니더
```suggestion
    .onDisappear() {
        // audio session 관련 메모리 해제 로직
        // 1. 플레이어 중지 및 메모리에서 해제
        highPlayer?.stop()
        lowPlayer?.stop()
        highPlayer = nil
        lowPlayer = nil
        
        // 2. 오디오 세션 비활성화
        do {
            try AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
        } catch {
            print("[Metronome] AudioSession deactivation error: \(error)")
        }
    }
```

ehdrbdndns commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,149 @@
+        //박자가 변경될 때 자연스럽게 CD 갯수가 바뀌는 애니메이션
+        .onChange(of: selectedTimeSig) { _, _ in
+            withAnimation(.spring(response: 0.28, dampingFraction: 0.85, blendDuration: 0.2)) {
+                if isPlaying { startDate = Date(); lastStepPlayed = -1 }
```
각 callback 함수 별로 함수 네이밍 지어주면 좋을듯? 근데 안해도 괜찮은 것 같아요

nakisara01 replied:
callback 부분은 일단 넘어가기로 합의했습니닷

ehdrbdndns reviewed (commented):
고생했셔유! 네이밍 정도만 수정하면 될듯?!

nakisara01 committed:
[fix] Apply review comment

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,164 @@
+        var body: some View {
+            HStack(spacing: beatsPerBar == 6 ? 18 : 24) {
+                ForEach(0 ..< beatsPerBar, id: \.self) { index in
+                    Image(index == currentBeat ? "metronome_cd_fill" : "metronome_cd_empty")
```
String으로 이미지 불러오지 않고 Xcode가 생성한 `ImageResource`를 사용하는 것이 좋을 것 같습니다.

```suggestion
                    Image(index == currentBeat ? .metronomeCdFill : .metronomeCdEmpty)
```

nakisara01 replied:
넵 변경된 metronome 에서 해당 부분 사용해서 적용했습니다.

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,164 @@
+    }
+
+    //오디오 출력 준비 (약간 audio play init 같은 느낌)
+    private func configureAudioSession() {
+        do {
+            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default, options: [.mixWithOthers])
+            try AVAudioSession.sharedInstance().setActive(true)
+        } catch {
+            print("[Metronome] AudioSession error: \(error)")
+        }
+    }
```
오디오 세션이 녹음 중에는 `.playAndRecord`여야 되어서 이 부분은 나중에 같이 논의해야 될 것 같습니다. (뷰 간의 오디오 세션 관리)

nakisara01 replied:
넵 일단 그러면 MARK로 표시해 두고 추후 다시 논의 하시죠

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,164 @@
+                    Image(index == currentBeat ? "metronome_cd_fill" : "metronome_cd_empty")
+                        .resizable()
+                        .scaledToFit()
+                        .frame(height: (beatsPerBar == 6 ? 33 : 44))
```
spacing이랑 height는 변수로 빼는 게 좋을 것 같습니다.

```suggestion
        private var spacing: CGFloat { beatsPerBar == 6 ? 18 : 24 }
        private var height: CGFloat { beatsPerBar == 6 ? 33 : 44 }
        
        var body: some View {
            HStack(spacing: spacing) {
                ForEach(0 ..< beatsPerBar, id: \.self) { index in
                    Image(index == currentBeat ? "metronome_cd_fill" : "metronome_cd_empty")
                        .resizable()
                        .scaledToFit()
                        .frame(height: height)
```

nakisara01 replied:
바뀐 디자인에도 수정해서 적용했습니다!

Segyun commented on GMG/GMG/UI/Components/MetronomeControlView.swift:
Metronome의 Preview로 옮겨도 될 것 같습니다!

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,164 @@
+    @State private var lowPlayer: AVAudioPlayer? = nil //메트로놈 약박 Player
+
+    // 박자에서 분자 뽑아내서 표시할 메트로놈 cd의 갯수 확정하기
+    private var beatsPerBar: Int { selectedTimeSig.numerator }
```
코드 전반에 걸쳐 마디를 나타내는 말을 Bar와 Measure 중에 하나로 통일하면 좋을 것 같습니다!

nakisara01 replied:
전체 다 Measure로 변경했습니다

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
Timer를 사용해서 수정해봤는데 TimelineView 사용하는 게 더 좋다는 이야기도 있네요. 다만 Timer로 사용하는게 더 직관적인 것 같습니다!

```Swift
import AVFoundation
import Combine
import SwiftUI

struct Metronome: View {
    let bpm: BPM
    let isPlaying: Bool
    let timeSignature: TimeSignature

    @State private var currentBeat: Int = 0
    @State private var timer: Publishers.Autoconnect<Timer.TimerPublisher>

    @State private var highPlayer: AVAudioPlayer? = nil  // 메트로놈 강박 Player
    @State private var lowPlayer: AVAudioPlayer? = nil  // 메트로놈 약박 Player

    // 한 마디 당 박의 수
    private var beatsPerMeasure: Int { timeSignature.numerator }

    init(
        bpm: BPM,
        isPlaying: Bool,
        timeSignature: TimeSignature
    ) {
        self.bpm = bpm
        self.isPlaying = isPlaying
        self.timeSignature = timeSignature

        self.timer = Timer.publish(
            every: bpm.beatDuration(),
            on: .main,
            in: .common
        )
        .autoconnect()
    }

    var body: some View {
        BeatBar(beatsPerBar: beatsPerMeasure, currentBeat: currentBeat)
            .padding()
            .onAppear {
                configureAudioSession()

                loadHighPlayer()
                loadLowPlayer()
            }
            .onDisappear {
                unloadHighPlayer()
                unloadLowPlayer()

                deactiveAudioSession()
            }
            .onReceive(timer) { _ in
                if isPlaying {
                    currentBeat = (currentBeat + 1) % beatsPerMeasure

                    playBeatSound(for: currentBeat)
                }
            }
            .onChange(of: bpm.value) { _, _ in
                timer.upstream.connect().cancel()
                timer = Timer.publish(
                    every: bpm.beatDuration(),
                    on: .main,
                    in: .common
                ).autoconnect()
            }
    }
}

// MARK: - Sub Views

extension Metronome {
    struct BeatBar: View {
        let beatsPerBar: Int
        let currentBeat: Int

        private var spacing: CGFloat { beatsPerBar == 6 ? 18 : 24 }
        private var height: CGFloat { beatsPerBar == 6 ? 33 : 44 }

        var body: some View {
            HStack(spacing: spacing) {
                ForEach(0..<beatsPerBar, id: \.self) { index in
                    Image(
                        index == currentBeat
                            ? "metronome_cd_fill" : "metronome_cd_empty"
                    )
                    .resizable()
                    .scaledToFit()
                    .frame(height: height)
                    .transition(
                        .asymmetric(
                            insertion: .scale.combined(with: .opacity),
                            removal: .opacity
                        )
                    )
                }
            }
            .animation(
                .spring(
                    response: 0.28,
                    dampingFraction: 0.85,
                    blendDuration: 0.2
                ),
                value: beatsPerBar
            )
            .animation(
                .spring(
                    response: 0.28,
                    dampingFraction: 0.85,
                    blendDuration: 0.2
                ),
                value: currentBeat
            )
        }
    }
}

// MARK: - Methods

extension Metronome {

    // MARK: Audio Session 관련 메서드

    // 오디오 출력 준비 (약간 audio play init 같은 느낌)
    private func configureAudioSession() {
        do {
            try AVAudioSession.sharedInstance().setCategory(
                .playback,
                mode: .default,
                options: [.mixWithOthers]
            )
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print("[Metronome] AudioSession error: \(error)")
        }
    }

    private func deactiveAudioSession() {
        do {
            try AVAudioSession.sharedInstance().setActive(
                false,
                options: .notifyOthersOnDeactivation
            )
        } catch {
            print(
                "[Metronome] AudioSession deactivation error: \(error)"
            )
        }
    }

    // MARK: Player 관련 메서드

    private func loadHighPlayer() {
        guard highPlayer == nil else { return }

        guard
            let url = Bundle.main.url(
                forResource: "Metronome_High",
                withExtension: "wav"
            )
        else {
            print(
                "[Metronome] Missing resource: Metronome_High.wav (check target membership)"
            )
            return
        }

        do {
            let player = try AVAudioPlayer(contentsOf: url)

            player.volume = 1.0
            player.prepareToPlay()

            self.highPlayer = player
        } catch {
            print("[Metronome] Failed to initialize audio player for high")
        }
    }

    private func loadLowPlayer() {
        guard lowPlayer == nil else { return }

        guard
            let url = Bundle.main.url(
                forResource: "Metronome_Low",
                withExtension: "wav"
            )
        else {
            print(
                "[Metronome] Missing resource: Metronome_Low.wav (check target membership)"
            )
            return
        }

        do {
            let player = try AVAudioPlayer(contentsOf: url)

            player.volume = 1.0
            player.prepareToPlay()

            self.lowPlayer = player
        } catch {
            print("[Metronome] Failed to initialize audio player for low")
        }
    }

    private func unloadHighPlayer() {
        guard let highPlayer else { return }
        highPlayer.stop()

        self.highPlayer = nil
    }

    private func unloadLowPlayer() {
        guard let lowPlayer else { return }
        lowPlayer.stop()

        self.lowPlayer = nil
    }

    // 소리 중 강박을 구분을 위한 인텍스 설정 (3/4, 4/4 박자는 첫 번째 박만 강박이지만, 6/8 박자의 경우 1, 4번 박에 강박을 줘야 함) *강박: Metronome 재생 중 기준을 잡는 높은 음
    private func shouldUseHigh(beatIndex: Int) -> Bool {
        if timeSignature.numerator == 6
            && timeSignature.denumerator == .eighth
        {
            return beatIndex == 0 || beatIndex == 3
        } else {
            return beatIndex == 0
        }
    }

    // 박자별 High or Low 음성 재생
    private func playBeatSound(for beatIndex: Int) {
        if shouldUseHigh(beatIndex: beatIndex) {
            guard let audioPlayer = highPlayer else {
                print("[Metronome] highPlayer is nil")
                return
            }
            audioPlayer.currentTime = 0
            audioPlayer.play()
        } else {
            guard let audioPlayer = lowPlayer else {
                print("[Metronome] lowPlayer is nil")
                return
            }
            audioPlayer.currentTime = 0
            audioPlayer.play()
        }
    }
}

#Preview {
    Metronome(bpm: BPM(value: 120), isPlaying: true, timeSignature: .fourFour)
}

// MARK: - BPM Extension

extension BPM {
    func beatDuration() -> TimeInterval {
        return 60.0 / TimeInterval(max(1, self.value))
    }
}
```

nakisara01 replied:
적용했습니다!

Segyun reviewed (changes requested):
수고하셨습니다~

나중에 로그 같은 경우에는 Core/Utils에 로거 함수 만들어서 사용하는 것도 좋을 것 같습니다!

nakisara01 committed:
[design] Apply modified metronome design

nakisara01 committed:
[fix] apply review content

 - TimeInterval -> TimelineView 기반의 뷰에서 Timer를 사용한 뷰로 변경했습니다.
- 디자인 시안 변경에 따른 Metronome 디자인을 변경했습니다
- CD 에셋을 변경했습니다
- TimeSignature struct에서 Equatable 설정을 빼고 extension을 선언해 적용했습니다.
-

nakisara01 committed:
[fix] metronome index init

- 메트로놈이 정지해있을 때 표시된 인덱스가 없도록 조정했습니다.

ehdrbdndns commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,269 @@
 struct Metronome: View {
+    let bpm: BPM //bpm
+    let isPlaying: Bool //메트로놈 플레이 여부
+    let timeSignature: TimeSignature //박자
```
```suggestion
    let bpm: BPM //bpm
    let isPlaying: Bool //메트로놈 플레이 여부
    let timeSignature: TimeSignature //박자
```

ㅋㅋㅋㅋ 위 내용들은 주석처리 안해도 잘 이해돼서 빼도 될듯해요!

nakisara01 replied:
넹 바꿨어영

ehdrbdndns reviewed (dismissed):
너무 수고 많으셨스요!

nakisara01 committed:
[fix] Delete MetronomeControlView & Add preview

- 시험용 UI였던 MetronomeContolView를 삭제하고 Metronome 내 Preview에 시험해볼 수 있는 형식을 추가했습니다.

nakisara01 committed:
[chore] delete trash annotation

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,308 @@
+        }
+    }
+
+    return PreviewWrapper()
```
`@Previewable` 쓰면 따로 뷰 안만들어도 돼요!

```suggestion
    @Previewable @State var bpm: Int = 100
    @Previewable @State var isPlaying: Bool = false
    @Previewable @State var timeSignature: TimeSignature = .fourFour

    VStack(spacing: 20) {
        Metronome(
            bpm: BPM(value: bpm),
            isPlaying: isPlaying,
            timeSignature: timeSignature
        )

        Button(isPlaying ? "Stop" : "Play") {
            isPlaying.toggle()
        }
        .buttonStyle(.bordered)

        Stepper("BPM: \(bpm)", value: $bpm, in: 30...320)

        HStack(spacing: 30) {
            Button("3/4") {
                timeSignature = .threeFour
            }
            .buttonStyle(.bordered)
            Button("4/4") {
                timeSignature = .fourFour
            }
            .buttonStyle(.bordered)
            Button("6/8") {
                timeSignature = .sixEight
            }
            .buttonStyle(.bordered)
        }
    }
    .padding()
```

nakisara01 replied:
오오 이런 방법이 있었군요! 감사합니다 수정할게요!!

Segyun commented on GMG/GMG/Resources/Assets.xcassets/MetronomeCD/MetronomeCD.imageset/Contents.json:
파일명도 이미지와 동일하게 수정하면 좋을 것 같습니다!

nakisara01 replied:
넵 MetronomeCD로 통일했습니다

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,308 @@
+    
     var body: some View {
-        Text(/*@START_MENU_TOKEN@*/"Hello, World!"/*@END_MENU_TOKEN@*/)
+        //TimelineView를 사용해 beatInterval마다 뷰가 업데이트 되도록 구성함
```
이전 설명 주석이 남아있어요.

nakisara01 replied:
삭제하겠습니다

Segyun commented on GMG/GMG/UI/Components/Metronome.swift:
```diff
@@ -6,13 +6,308 @@
+        private var width: CGFloat { beatsPerMeasure == 6 ? 34 : 46 }
+        
+        var body: some View {
+            Image(.metronomeCD)
```
Assets에는 `MetronomeCd`라고 되어있어서 없다고 나와요. 에셋의 이미지명을 수정해야 할 듯합니다!

nakisara01 replied:
파일을 MetronomeCD로 수정했습니다!

nakisara01 committed:
[fix] Apply review content

- Preview에서 상태값에 Previewable를 달아 View를 사용하지 않고도 preview를 사용할 수 있도록 조정했습니다
- MetronomeCD의 파일명을 통일했습니다.

Segyun reviewed (approved):
LGTM!
